presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java

```diff
@@ -0,0 +1,86 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.execution.QueryManager;
+import com.facebook.presto.execution.resourceGroups.ResourceGroupManager;
+import com.facebook.presto.spi.resourceGroups.ResourceGroupId;
+import com.facebook.presto.spi.resourceGroups.ResourceGroupInfo;
+import com.google.common.collect.ImmutableList;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.server.QueryStateInfo.createQueryStateInfo;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.toSet;
+
+@Path("/v1/query_state")
+public class QueryStateInfoResource
+{
+    private final QueryManager queryManager;
+    private final ResourceGroupManager resourceGroupManager;
+
+    @Inject
+    public QueryStateInfoResource(
+            QueryManager queryManager,
+            ResourceGroupManager resourceGroupManager)
+    {
+        this.queryManager = requireNonNull(queryManager, "queryManager is null");
+        this.resourceGroupManager = requireNonNull(resourceGroupManager, "resourceGroupManager is null");
+    }
+
+    @GET
+    public List<QueryStateInfo> getQueryDiagnosticsByUser(@QueryParam("user") String user)
+    {
+        if (user == null || user.isEmpty()) {
+            return emptyList();
+        }
+
+        List<QueryInfo> userQueries = queryManager.getAllQueryInfo().stream()
+                .filter(queryInfo -> user.equals(queryInfo.getSession().getUser()))
+                .collect(toImmutableList());
+        Set<ResourceGroupId> resourceGroupIdSet = userQueries.stream()
+                .map(queryInfo -> queryManager.getQueryResourceGroup(queryInfo.getQueryId()))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(toSet());
+        Map<ResourceGroupId, ResourceGroupInfo> rootResourceGroupInfos = resourceGroupIdSet.stream().collect(toMap(identity(), id -> resourceGroupManager.getResourceGroupInfo(id.getRoot())));
```

Comment:
This seems to construct the tree once for every leaf resource group, which is horribly inefficient.